README.md
# react-native-dots-pagination

Paging as dots for react-native.

<p align="center">
  <img src="https://raw.githubusercontent.com/tsepeti/react-native-dots-pagination/master/record.gif" width="50%" />
</p>


<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->


  - [Quickstart](#quickstart)
  - [Using](#using)
- [Props](#props)
- [Issues](#issues)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Quickstart
    yarn add react-native-dots-pagination
    
## Using

```JS
import React, { Component } from 'react';
import Dots from 'react-native-dots-pagination';


class Example extends Component {
  constructor(props) {
    super(props);
    this.state = {
      active: 0
    }
  }
  
  render() {
    return (
      <Dots length={10} active={this.state.active} />
    )
  }
}
```

# Props

| Name | Default value | Description |
|--|--|--|
| length | 10 | Required. The amount of dots you want to use. |
| active | 1 | Required. The index of the currently active dot. |
| width | 300 | The overall width works like a center. |
| paddingVertical | 10 | Average Vertical padding. |
| paddingHorizontal | 10 | Average Horizontal padding. |
| passiveDotWidth | 10 | (Width, Height) for passive dot. |
| activeDotWidth | 15 | (Width, Height) for active dot. |
| passiveDotHeight | 10 | Height for passive dot. |
| activeDotHeight | 15 | Height for active dot. |
| passiveColor | #CCCCCC | Colors for passive dots. |
| activeColor | #016bd8 | Colors for active dots. |

# Issues

[Github Issues](https://github.com/tsepeti/react-native-dots-pagination/issues) are used to track todos, bugs, feature requests, and more.



History.md
# v.NEXT

## 0.1.10
* Fixed full width center. Thanks [@mstrk](https://github.com/mstrk)

## 0.1.9
* Update to follow React-Native best practices (componentWillReceiveProps) #3
* Fixed ref attributes is deprecated (react/no-string-refs)

## 0.1.8
* added Allow to set specific height for dots, Thanks [@vdrorofeev](https://github.com/vdrorofeev)

## 0.1.7
* Align dots on x axis thanks [@dbarner1](https://github.com/dbarner1)

## 0.1.6
* fixed dots width

## 0.1.5
* prop-types package added.

## 0.1.4
* fix android showsHorizontalScrollIndicator

## 0.1.3
* added active dot border.

## 0.1.2
* PropTypes isRequired params.

## 0.1.1
* android toScroll function fix.

## 0.1.0
* First Release


example/App.js
import React, {Component} from 'react';
import {
  SafeAreaView,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';

import Dots from 'react-native-dots-pagination';
import _ from 'underscore';

class Example extends Component {
  state = {
    active0: 0,
    active1: 1,
    active2: 2,
    active3: 3,
  };

  changePress() {
    this.setState({
      active0: _.random(0, 5),
      active1: _.random(0, 10),
      active2: _.random(0, 15),
      active3: _.random(0, 20),
    });
  }

  render() {
    return (
      <SafeAreaView>
          <Dots length={5} active={this.state.active0} />
          <Dots length={10} active={this.state.active1} />
          <Dots length={15} active={this.state.active2} />
          <Dots length={20} active={this.state.active3} />

        <View style={Styles.footer}>
          <TouchableOpacity
            style={Styles.button}
            onPress={this.changePress.bind(this)}>
            <Text style={Styles.buttonText}>CHANGE</Text>
          </TouchableOpacity>
        </View>
      </SafeAreaView>
    );
  }
}

const Styles = StyleSheet.create({
  footer: {
    alignItems: 'center',
    marginVertical: 15,
  },

  button: {
    width: 100,
    padding: 10,
    backgroundColor: '#333',
    borderRadius: 5,
    alignItems: 'center',
  },

  buttonText: {
    color: 'white',
    fontWeight: 'bold',
  },
});

export default Example;


index.js
import React, { Component } from 'react';
import { View, ScrollView, Platform } from 'react-native';
import PropTypes from 'prop-types';

// Styles
import Styles from './styles';

// set to fit
const scalesPageToFit = Platform.OS === 'android';

const DEFAULT_PASSIVE_DOT_WIDTH = 10;
const DEFAULT_ACTIVE_DOT_WIDTH = 15;

export default class Dots extends Component {
  static get propTypes() {
    return {
      length: PropTypes.number.isRequired,
      active: PropTypes.number.isRequired,

      width: PropTypes.number,
      paddingVertical: PropTypes.number,
      paddingHorizontal: PropTypes.number,

      // Dots
      passiveDotWidth: PropTypes.number,
      activeDotWidth: PropTypes.number,
      activeDotHeight: PropTypes.number,
      passiveDotHeight: PropTypes.number,
      passiveColor: PropTypes.string,
      activeColor: PropTypes.string,

      // active Border
      activeBorder: PropTypes.bool,
      activeBorderColor: PropTypes.string,
      activeBorderWidth: PropTypes.number,
    };
  }

  static defaultProps = {
    width: 300,

    marginHorizontal: 2,
    paddingVertical: 10,
    paddingHorizontal: 10,

    passiveDotWidth: DEFAULT_PASSIVE_DOT_WIDTH,
    activeDotWidth: DEFAULT_ACTIVE_DOT_WIDTH,
    activeDotHeight: DEFAULT_ACTIVE_DOT_WIDTH,
    passiveDotHeight: DEFAULT_PASSIVE_DOT_WIDTH,
    passiveColor: '#CCCCCC',
    activeColor: '#016bd8',

    activeBorder: false,
    activeBorderWidth: 3,
    activeBorderColor: '#FFF',
  };

  componentDidUpdate (prevProps) {
    const newActive = this.props.active;
    if (prevProps.active !== newActive) {
      this.scrollTo(newActive);
    }
  }

  scrollTo(index) {
    const { width, activeDotWidth } = this.props;
    const get = this[`dots_${index}`];

    if (get) {
      const x = get.x - (width / 2 - activeDotWidth);

      return this.scrollRef.scrollTo({ x });
    }
  }

  isActive(index) {
    return this.props.active === index;
  }

  dotStyle(isActive) {
    const {
      activeDotWidth,
      passiveDotWidth,
      activeDotHeight,
      passiveDotHeight,
      activeColor,
      passiveColor,
      activeBorder,
      activeBorderWidth,
      activeBorderColor,
      marginHorizontal,
      alignDotsOnXAxis,
    } = this.props;
    const width = isActive ? activeDotWidth : passiveDotWidth;
    const marginTop = alignDotsOnXAxis || !isActive ? 0 : -width / 6;

    let height = width;

    if (isActive && activeDotHeight != null) {
      height = activeDotHeight;
    } else if (!isActive && passiveDotHeight != null) {
      height = passiveDotHeight;
    }

    let style = {
      width,
      height,
      marginHorizontal,
      backgroundColor: isActive ? activeColor : passiveColor,
      borderRadius: width,
      marginTop,
    };

    // active Border Styles.
    if (activeBorder && isActive) {
      style.borderWidth = activeBorderWidth;
      style.borderColor = activeBorderColor;
    }

    return style;
  }

  render() {
    const {
      length,
      width,
      paddingVertical,
      paddingHorizontal,
      activeDotWidth,
      activeBorderWidth,
      passiveDotWidth,
      marginHorizontal,
    } = this.props;
    const list = [...Array(length).keys()];
    const activeWidth = (activeBorderWidth * 4) + activeDotWidth + paddingHorizontal;
    const scrollWidth = activeWidth + ((list.length - 1) * passiveDotWidth) + (marginHorizontal * (list.length * 2));

    return (
      <View style={Styles.container}>
        <ScrollView
          ref={(el) => { this.scrollRef = el; }}
          style={{ width: width < scrollWidth ? width : scrollWidth }}
          contentContainerStyle={{ alignItems: 'center', paddingVertical, paddingHorizontal }}
          scalesPageToFit={scalesPageToFit}
          bounces={false}
          horizontal={true}
          scrollEnabled={false}
          showsHorizontalScrollIndicator={false}
        >
          {list.map(i => {
            return (
              <View
                key={i}
                style={this.dotStyle(this.isActive(i))}
                onLayout={({
                  nativeEvent: {
                    layout: { x, y },
                  },
                }) => {
                  this[`dots_${i}`] = { x, y };
                }}
              />
            );
          })}
        </ScrollView>
      </View>
    );
  }
}
